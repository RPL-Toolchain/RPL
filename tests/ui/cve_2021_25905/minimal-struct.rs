//@ revisions: inline regular
//@[inline] compile-flags: -Z inline-mir=true
//@[inline] ignore-on-host
//@[regular] compile-flags: -Z inline-mir=false

#[derive(Default)]
pub struct Wrapper {
    buf: Vec<u8>,
}

// #[rpl::dump_mir(dump_cfg, dump_ddg)]
fn main() {
    let mut wrapped: Wrapper = Wrapper::default();
    let b = wrapped.buf.len();
    let buf = unsafe {
        std::slice::from_raw_parts_mut(wrapped.buf.as_mut_ptr().offset(b as isize), wrapped.buf.capacity() - b)
        //~^ ERROR: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
    };
}
