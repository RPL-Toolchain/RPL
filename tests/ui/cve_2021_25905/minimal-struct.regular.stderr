error: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
  --> tests/ui/cve_2021_25905/minimal-struct.rs:15:9
   |
LL |     let b = wrapped.buf.len();
   |             ----------------- slice created with this length
LL |     let buf = unsafe {
LL |         std::slice::from_raw_parts_mut(wrapped.buf.as_mut_ptr().offset(b as isize), wrapped.buf.capacity() - b)
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |                              |
   |         |                              slice created with this pointer
   |         slice created here
   |
   = help: See https://doc.rust-lang.org/std/slice/fn.std::slice::from_raw_parts_mut.html
   = note: `#[deny(rpl::slice_from_raw_parts_uninitialized)]` on by default

error: aborting due to 1 previous error

